src/heroku.js

```diff
@@ -0,0 +1,197 @@
+  return { email: HEROKU_EMAIL, apiKey: HEROKU_API_KEY };
+};
+
+///
+/// Heroku API read functions
+///
+
+/* Loads information about the given app */
+module.exports.getApp = async function (appName) {
+  return herokuGet(`https://api.heroku.com/apps/${appName}`);
+};
+
+/* Loads the UUID of the named pipeline from Heroku. */
+module.exports.getPipelineId = async function (pipelineName) {
+  try {
+    const data = await herokuGet(`https://api.heroku.com/pipelines/${pipelineName}`);
+    return data.id;
+  } catch (err) {
+    if (err instanceof HttpResponseError && err.status === 404) {
+      return undefined;
+    }
+    throw err;
+  }
+};
+
+/* Loads all of the config vars for the given pipeline's "Review Apps" stage. */
+module.exports.getPipelineVars = async function (pipelineId) {
+  return herokuGet(`https://api.heroku.com/pipelines/${pipelineId}/stage/review/config-vars`);
+};
+
+/*
+ * Loads the review app configuration for the given pipeline. Returns
+ * undefined if the pipeline does not have review apps configured.
+ */
+module.exports.getReviewAppConfig = async function (pipelineId) {
+  try {
+    const resp = await herokuGet(`https://api.heroku.com/pipelines/${pipelineId}/review-app-config`);
+    return resp;
+  } catch (err) {
+    if (err instanceof HttpResponseError && err.status === 404) {
+      return undefined;
+    }
+    throw err;
+  }
+};
+
+/* Checks whether an app with the given name exists on Heroku. Returns bool. */
+module.exports.appExists = async function (appName) {
+  try {
+    await herokuFetch(`https://api.heroku.com/apps/${appName}`, { method: 'GET' });
+    return true;
+  } catch (err) {
+    if (err instanceof HttpResponseError && err.status === 404) {
+      return false;
+    }
+    throw err;
+  }
+};
+
+///
+/// Heroku API write functions
+///
+
+/* Creates a Heroku app with the given name. */
+module.exports.createApp = async function (name) {
+  const resp = await herokuFetch('https://api.heroku.com/teams/apps', {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
+    body: JSON.stringify({ name, region: 'us', team: 'readme' }),
+  });
+  return resp.json();
+};
+
+/* Deletes a Heroku app with the given name. */
+module.exports.deleteApp = async function (name) {
+  if (!/-pr-/.test(name)) {
+    throw new Error('This function can only delete pull request review apps');
+  }
+  const resp = await herokuFetch(`https://api.heroku.com/apps/${name}`, {
+    method: 'DELETE',
+    headers: { 'Content-Type': 'application/json' },
+  });
+  return resp.json();
+};
+
+/*
+ * Attaches the given app to the given pipeline. Uses the "development" stage
+ * because apps won't appear on the Heroku Dashboard if they're in "review"
+ * stage and the pipeline isn't connected to GitHub.
+ */
+module.exports.coupleAppToPipeline = async function (appId, pipelineId) {
+  const resp = await herokuFetch('https://api.heroku.com/pipeline-couplings', {
```

Comment:
Is the coupling functionality what triggers a rebuild?

Reply:
Nope, a "pipeline coupling" is the object that establishes the relationship between an app and a pipeline. We create the app, then we create a coupling so that the app appears in the correct pipeline, rather than in our top-level list of apps.

More details in Heroku docs: https://devcenter.heroku.com/articles/pipelines-using-the-platform-api#adding-apps-to-a-pipeline

(Builds are triggered by a `git push`, which happens in `src/git.js`.)